update pyramid's default execution policy to attempt to render any uncaught exception

If an exception propagates all the way to the execution policy there should be a last-ditch effort to render a response... this is part of the larger consistency change to pyramid_tm and pyramid_retry. With the addition of `request.invoke_exception_view` it's hopefully clear that exception views should be callable from various parts of the lifecycle and pyramid_retry will attempt to invoke exception views at the end if an exception is uncaught. Also pyramid_tm will attempt to invoke exceptions on commit/abort to handle those specifically.

I've documented the process here in a little ascii diagram[1] reproduced here:
```
   pyramid_retry   +-------->    if exc not retryable, render response
                                 if exc retryable, ignore exc and try again
        +
        |
        |
        v

pyramid_debugtoolbar

        +
        |
        |
        v

    pyramid_tm     +-------->    if exc caught then abort and reraise
                                 if request.exception then abort and return response
        +                        if abort or commit raises, render response
        |
        |
        v

     excview       +-------->    if exc caught, render response and set request.exception
                                                if no exception then reraise
        +
        |
        |
        v

       app
```

[1] https://gist.github.com/mmerickel/2e83f4af6c7d5eae34947b8f1f075f61

```
    def test_it_supports_alternate_requests(self):
        def exc_view(exc, request):
            self.assertTrue(request is other_req)

# Next two lines are new
            from pyramid.threadlocal import get_current_request
            self.assertTrue(get_current_request() is other_req)

            return DummyResponse(b'foo')
        self.config.add_view(exc_view, context=RuntimeError)
        request = self._makeOne()
        other_req = self._makeOne()
        try:
            raise RuntimeError
        except RuntimeError:
            response = request.invoke_exception_view(request=other_req)
            self.assertEqual(response.app_iter, [b'foo'])
        else: # pragma: no cover
            self.fail()
```

This test, changed from the original in `test_view.py` will fail:

```
======================================================================
FAIL: test_it_supports_alternate_requests (pyramid.tests.test_view.TestViewMethodsMixin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xistence/Projects/pyramid/pyramid/tests/test_view.py", line 800, in test_it_supports_alternate_requests
    raise RuntimeError
RuntimeError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xistence/Projects/pyramid/pyramid/tests/test_view.py", line 802, in test_it_supports_alternate_requests
    response = request.invoke_exception_view(request=other_req)
  File "/Users/xistence/Projects/pyramid/pyramid/view.py", line 702, in invoke_exception_view
    request_iface=request_iface.combined,
  File "/Users/xistence/Projects/pyramid/pyramid/view.py", line 612, in _call_view
    response = view_callable(context, request)
  File "/Users/xistence/Projects/pyramid/pyramid/viewderivers.py", line 409, in viewresult_to_response
    result = view(context, request)
  File "/Users/xistence/Projects/pyramid/pyramid/tests/test_view.py", line 794, in exc_view
    self.assertTrue(get_current_request() is other_req)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 2612 tests in 8.742s
```

This is due to the fact that the threadlocals are not pushed when invoking the exception view, this means that the `request` in the view is different and potentially has different behaviour than the `get_current_request()` `request` object.

--- a/pyramid/router.py
+++ b/pyramid/router.py
@@ -1,3 +1,4 @@
+import sys
 from zope.interface import (
     implementer,
     providedBy,
@@ -24,6 +25,7 @@ from pyramid.events import (
     BeforeTraversal,
     )
 
+from pyramid.compat import reraise
 from pyramid.httpexceptions import HTTPNotFound
 from pyramid.request import Request
 from pyramid.view import _call_view
@@ -252,7 +254,15 @@ class Router(object):
         response = self.execution_policy(environ, self)
         return response(environ, start_response)
 
-
 def default_execution_policy(environ, router):
     request = router.make_request(environ)
-    return router.invoke_request(request)
+    try:
+        return router.invoke_request(request)
+    except Exception:
+        exc_info = sys.exc_info()
+        try:
+            return request.invoke_exception_view(exc_info)
+        except HTTPNotFound:
+            reraise(*exc_info)
+        finally:
+            del exc_info  # avoid local ref cycle

--- a/pyramid/tests/test_router.py
+++ b/pyramid/tests/test_router.py
@@ -1284,6 +1284,33 @@ class TestRouter(unittest.TestCase):
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.body, b'foo')
 
+    def test_execution_policy_handles_exception(self):
+        from pyramid.interfaces import IViewClassifier
+        from pyramid.interfaces import IExceptionViewClassifier
+        from pyramid.interfaces import IRequest
+        class Exception1(Exception):
+            pass
+        class Exception2(Exception):
+            pass
+        req_iface = self._registerRouteRequest('foo')
+        self._connectRoute('foo', 'archives/:action/:article', None)
+        view = DummyView(DummyResponse(), raise_exception=Exception1)
+        self._registerView(view, '', IViewClassifier, req_iface, None)
+        exception_view1 = DummyView(DummyResponse(),
+                                    raise_exception=Exception2)
+        self._registerView(exception_view1, '', IExceptionViewClassifier,
+                           IRequest, Exception1)
+        response = DummyResponse()
+        response.app_iter = ["Hello, world"]
+        exception_view2 = DummyView(response)
+        self._registerView(exception_view2, '', IExceptionViewClassifier,
+                           IRequest, Exception2)
+        environ = self._makeEnviron(PATH_INFO='/archives/action1/article1')
+        start_response = DummyStartResponse()
+        router = self._makeOne()
+        result = router(environ, start_response)
+        self.assertEqual(result, ["Hello, world"])
+
 class DummyPredicate(object):
     def __call__(self, info, request):
         return True

